### PR TITLE
Work Item Write Operations (Create/Update)

### PR DESCRIPTION
**Pull request by gkillough**

### CODE REVIEW

File: azure-boards-common/src/main/java/com/synopsys/integration/azure/boards/common/http/AzureHttpService.java [hunk 1]
```diff
@@ -47,29 +44,27 @@
...
-        return withCredential(baseUrl, Proxy.NO_PROXY, oAuthCredential, gson);
-    }
-
-    public static AzureHttpService withCredential(String baseUrl, Proxy proxy, Credential oAuthCredential, Gson gson) {
-        NetHttpTransport netHttpTransport = new NetHttpTransport.Builder()
-                                                .setProxy(proxy)
-                                                .build();
-        return new AzureHttpService(baseUrl, netHttpTransport.createRequestFactory(oAuthCredential), gson);
-    }
-
```
Comment (gkillough) [inline azure-boards-common/src/main/java/com/synopsys/integration/azure/boards/common/http/AzureHttpService.java:60]: Moved to factory class.

File: azure-boards-common/src/main/java/com/synopsys/integration/azure/boards/common/http/AzureHttpServiceFactory.java [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+
+    public static AzureHttpService withCredentialNoProxy(Credential oAuthCredential, Gson gson) {
+        return withCredential(DEFAULT_BASE_URL, Proxy.NO_PROXY, oAuthCredential, gson);
+    }
+
+    public static AzureHttpService withCredentialNoProxy(String baseUrl, Credential oAuthCredential, Gson gson) {
+        return withCredential(baseUrl, Proxy.NO_PROXY, oAuthCredential, gson);
+    }
+
+    public static AzureHttpService withCredential(Proxy proxy, Credential oAuthCredential, Gson gson) {
```
Comment (bamandel) [inline azure-boards-common/src/main/java/com/synopsys/integration/azure/boards/common/http/AzureHttpServiceFactory.java:42]: Looks like this method can call the method below it.

File: azure-boards-common/src/main/java/com/synopsys/integration/azure/boards/common/service/workitem/request/WorkItemElementOperationModel.java [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+
+public class WorkItemElementOperationModel {
+    private String op;
+    private String path;
+    private String from;
+    private Object value;
+
+    public static WorkItemElementOperationModel fieldElement(WorkItemElementOperation operation, AzureFieldDefinition<?> fieldDefinition, Object value) {
+        String fieldPath = String.format("/fields/%s", fieldDefinition.getFieldName());
+        return new WorkItemElementOperationModel(operation.toLowerCaseString(), null, fieldPath, value);
```
Comment (bamandel) [inline azure-boards-common/src/main/java/com/synopsys/integration/azure/boards/common/service/workitem/request/WorkItemElementOperationModel.java:35]: The name of fieldPath may be confusing me, but are some variables switched here? Should null be the third param?

File: azure-boards-common/src/main/java/com/synopsys/integration/azure/boards/common/service/workitem/AzureWorkItemService.java [hunk 4]
```diff
@@ -44,4 +53,39 @@ public WorkItemResponseModel getWorkItem(String organizationName, String project
...
+    }
+
+    public WorkItemResponseModel updateWorkItem(String organizationName, String projectIdOrName, Integer workItemId, WorkItemRequest workItemRequest) throws HttpServiceException {
+        String requestSpec = API_SPEC_ORGANIZATION_PROJECT_WORKITEMS_INDIVIDUAL
+                                 .defineReplacement("{organization}", organizationName)
+                                 .defineReplacement("{project}", projectIdOrName)
+                                 .defineReplacement("{workitemId}", workItemId.toString())
+                                 .populateSpec();
+        try {
+            HttpRequest httpRequest = buildWriteRequest(HttpMethods.PATCH, requestSpec, workItemRequest.getElementOperationModels());
```
Comment (jamesrichard91) [inline azure-boards-common/src/main/java/com/synopsys/integration/azure/boards/common/service/workitem/AzureWorkItemService.java:77]: Is PATCH the right Http method here? Are you sure it shouldn't be PUT?

Comment (gkillough) [reply azure-boards-common/src/main/java/com/synopsys/integration/azure/boards/common/service/workitem/AzureWorkItemService.java:77]: It is strange, but they don't support `PUT`. Their documentation for "Update" says to use `PATCH` and just change the fields you need to.

File: azure-boards-common/src/main/java/com/synopsys/integration/azure/boards/common/service/workitem/request/WorkItemElementOperationModel.java (hunk 3 above)
Comment (gkillough) [reply azure-boards-common/src/main/java/com/synopsys/integration/azure/boards/common/service/workitem/request/WorkItemElementOperationModel.java:35]: Good catch.

**Pull request merged**